src/assets/scripts/client/trafficGenerator/buildPreSpawnAircraft.js

```diff
@@ -1,78 +1,245 @@
+ *
+ * @function _calculateOffsetsToEachWaypointInRoute
+ * @param waypointModelList {array<WaypointModel>} array of all waypoints along the route
+ * @return {array<number>} offset distance from spawn point to each waypoint
+ */
+function _calculateOffsetsToEachWaypointInRoute(waypointModelList) {
+    // begin by storing the first fix's offset: a distance of 0 from the spawn point
+    const waypointOffsetMap = [0];
+
+    // continue with second waypoint, because first is already stored
+    for (let i = 1; i < waypointModelList.length; i++) {
+        if (waypointModelList[i].isVectorWaypoint || waypointModelList[i - 1].isVectorWaypoint) {
+            continue;
+        }
+
+        const previousWaypointModel = waypointModelList[i - 1];
+        const nextWaypointModel = waypointModelList[i];
+        const distanceToNextWaypoint = previousWaypointModel.calculateDistanceToWaypoint(nextWaypointModel);
+
+        waypointOffsetMap.push(distanceToNextWaypoint);
+    }
+
+    return waypointOffsetMap;
+}
+
+/**
+ * Create an array of the various altitude restrictions along the route, and the offset distances at which they exist
+ *
+ * Note that unlike the `waypointOffsetMap`, the altitude offset array DOES NOT mirror the indices
+ * of the `waypointModelList`. Only offsets and altitudes of altitude-restricted waypoints are included.
+ *
+ * @function _calculateAltitudeOffsets
+ * @param waypointModelList {array<WaypointModel>} array of all waypoints along the route
+ * @param waypointOffsetMap {array<number>} offset distance from spawn point to each waypoint
+ * @return {array<array<number>>} [[offsetDistance, altitude], [offsetDistance, altitude], ...]
+ */
+function _calculateAltitudeOffsets(waypointModelList, waypointOffsetMap) {
+    const altitudeOffsets = [];
+
+    _forEach(waypointModelList, (waypointModel, waypointIndex) => {
+        if (!waypointModel.hasAltitudeRestrictions) {
+            return;
+        }
+
+        altitudeOffsets.push([
+            waypointOffsetMap[waypointIndex],
+            Math.min(waypointModel.altitudeMaximum, waypointModel.altitudeMinimum)
+        ]);
+    });
+
+    return altitudeOffsets;
+}
+
+/**
+ * Calculate the ideal spawn altitude at the given distance along the route from the spawn point
+ *
+ * @function _calculateAltitudeAtOffset
+ * @param waypointModelList {array<WaypointModel>} array of all waypoints along the route
+ * @param waypointOffsetMap {array<number>} offset distance from spawn point to each waypoint
+ * @param altitudeOffsets {array<array<number>>} information about location of altitude restrictions
+ * @param offsetDistance {number} distance along route at which we want to know the ideal spawn altitude
+ * @return {number} ideal spawn altitude, in feet (rounded up to nearest thousand)
+ */
+function _calculateAltitudeAtOffset(waypointModelList, waypointOffsetMap, altitudeOffsets, offsetDistance) {
```

Comment:
I feel like this method should be split up. There is a lot of motive here and this is a pretty long method. Not something that needs the happen now, just thought I’d call it out.

Reply:
I agree it feels heavy. The iffing is needed unfortunately, and though we could do their work with more helper functions, the logic will never be used elsewhere anyway, so I'm inclined to skip pulling them out in this case if that's cool.